cmd/gcp-controller-manager/app/ca_cache_test.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"math/big"
+	"net/http"
+	"net/http/httptest"
+	"path"
+	"testing"
+	"time"
+)
+
+func TestCACacheVerify(t *testing.T) {
+	var ca fakeCA
+	srv := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		switch path.Base(r.URL.Path) {
+		case "root.crt":
+			rw.Write(ca.rootCert)
+		case "intermediate.crt":
+			rw.Write(ca.intermediateCert)
+		case "root.crl":
+			rw.Write(ca.rootCRL)
+		case "intermediate.crl":
+			rw.Write(ca.intermediateCRL)
+		default:
+			http.Error(rw, "not found", http.StatusNotFound)
+		}
+	}))
+	defer srv.Close()
+	ca = initFakeCA(t, srv.URL)
+
+	c := &caCache{
+		rootCertURL: srv.URL + "/root.crt",
+		interPrefix: srv.URL,
+		certs:       make(map[string]*x509.Certificate),
+		crls:        make(map[string]*cachedCRL),
+	}
+
+	if err := c.verify(ca.signedCert); err != nil {
+		t.Errorf("verifying valid certificate: got %v, want nil", err)
+	}
+	if err := c.verify(ca.selfSignedCert); err == nil {
+		t.Errorf("verifying self-signed certificate: got nil, want non-nil error")
+	}
+	if err := c.verify(ca.revokedCert); err == nil {
+		t.Errorf("verifying revoked certificate: got nil, want non-nil error")
+	}
+	if err := c.verify(ca.noCRLCert); err == nil {
+		t.Errorf("verifying certificate without CRLs: got nil, want non-nil error")
+	}
+}
```

Comment:
Maybe add a test for a leaf cert that points to an invalid intermediate (and maybe have the intermediate point to an invalid root as well?)

Reply:
Done